### DISCUSSION TITLE
How to scale my 4K game to play in 1080p?

### DISCUSSION THREAD

**Discussion by WithinAmnesia** [flareteam/flare-game#937] [Q&A] [answered]
https://github.com/flareteam/flare-game/discussions/935
Currently my video game runs on 4k and I'm not sure how to down scale the screen to run on 1080p with all the dimensions intact? I'm hoping to get 1080p support for most people play on 1080p machines currently today mhmm. How can this be done hey?

**Comment by dorkster** [comment, marked as answer]
The engine will scale down the render output to fit the window. In `engine/resolutions.txt`, lower the `required_width` and `required_height` values to set your minimum window size.

![Screenshot_20230302_224142](https://user-images.githubusercontent.com/913982/222626066-c2b182e9-fc73-493d-80e3-ada18a26209a.jpg)

As you can see, the UI gets scaled down as well. So you'll need to keep things like text size in mind.

**Reply by WithinAmnesia** [reply]
So this works on my 1080 windows laptop and yes I need to fix the font size to be readable:
"# Screen and icon resolution settings

menu_frame_width=640
menu_frame_height=480

icon_size=128

required_width=1920
required_height=1080

virtual_height=2160
virtual_dpi=91.754738

#3840 	× 	2160"

However the player character and zombies are not loading like they are invisible (yet the skeletons are loading in) yet they work and attack and make noises yet the graphics are not loading in.

@dorkster Any ideas on what this could be? How can this issue be fixed?

**Reply by dorkster** [reply]
No clue. Maybe try switching the renderer in Flare's video settings to `sdl` and see if the issue persists? With a laptop GPU, you might be hitting a texture size limit.

**Comment by WithinAmnesia** [comment]
![1080p screen resolution version of my video game working!](https://user-images.githubusercontent.com/86871191/222919023-cf48663b-2e6d-4031-8416-8c0d9d640062.png)
1080p version of my video game working! The sdl rendering seemed to work great! Also before and during the successful test I have the .exe on run as administrator and also high dpi over override on application. The 2018 ish 1k 1080p gaming laptop was able to chug along at 16-7 fps, like not great. I still need to fix some gui issues like make the text and menus bigger so they show up good in 4K and 1080p. Also I wonder how to set up Flare where the dpi and sdl settings are to have the highest chance of successful first launch in how this can be achieved? The low frame rate but the game running fine mechanically reminded myself of old school gaming on big clunky 90's pc's and flash games chugging along on old pc's. So hardware will get better over time too so like if a ~5 year old like kinda crappy gaming laptop can still chug along I'm pretty sure as this ages that the hardware demands will be less of an issue mhmm. People's computers are getting better over time and other older pc's eventually get better one's to replace them. So I feel having a slightly demanding game is not the bad for I want to build a game to last anyway and lots of games get dated too quickly catering to yesteryear's technology when they launch. the 90's-2000's games were sorta bad for that. So I hope to try and be forward thinking so 10-20 years from now my video game still looks beautiful and is worth playing; built to last.

**Comment by WithinAmnesia** [comment]
@dorkster How can one scale / adjust the exit menu? It was deleted / removed in 1.14 I think I'm not really finding what the 1.14 version uses for an exit menu, it has tabs. I'm trying to make the menus 4 times the size and so far I have the font working more or less and the buttons but I cannot figure out currently how to adjust the exit menu? I'm just trying to scale everything up by 4 times so when the 4k version goes to scale on 1080p it still looks excellent.

**Reply by dorkster** [reply]
The config/exit menu is defined at [menus/config.txt](https://github.com/flareteam/flare-engine/blob/master/mods/default/menus/config.txt), which ships in the "default" mod with the engine. The background image is located at [images/menus/config.png](https://github.com/flareteam/flare-game/blob/master/mods/fantasycore/images/menus/config.png).

**Reply by WithinAmnesia** [reply]
@dorkster How can one fix this so that it looks like a regular config menu but 4 times larger?

The images and fonts are scaled 4 times larger. '+' = # so the comments works.
![image](https://user-images.githubusercontent.com/86871191/222929641-b6268ab5-4204-488d-8924-af4fb66ca00c.png)
![button_default](https://user-images.githubusercontent.com/86871191/222929687-72b2e3ad-e534-4fc3-9d62-e82bf53bf5a8.png)
![config](https://user-images.githubusercontent.com/86871191/222929690-6997c8e0-a9ba-445c-9a9a-a41c5a2ca0f8.png)

config.txt
"'+'position=-600,-600
'+'align=topleft


'+' Configuration menu Widgets Definitions.
'+' Coordinates should be integer, representing absolute position on 640x480 screen
'+' First two values are for label position (x,y) and two last for control (x,y)
'+' Buttons don't have separate labels

button_ok=-0,-56,bottom,center,top
button_defaults=0,-28,bottom,center,top
button_cancel=0,0,bottom,center,top

listbox_scrollbar_offset=2,center,top

'+' scrollpane values are positions x, y and sizes width, height
'+' scrollpane=0,4,600,320
scrollpane=-600,-600,2464,1376,center,top
scrollpane_padding=320,160,center,top
'+' scrollpane_bg_color=26,26,26,127
scrollpane_separator_color=63,63,63

[mods]
activemods=456,16,384,32
inactivemods=136,16,64,32
activemods_shiftup=544,128
activemods_shiftdown=544,160
activemods_deactivate=224,160
inactivemods_activate=224,128"


font_settings.txt
"'+' Only .ttf files are supported, due to limitations in SDL_ttf
'+' The font must be placed in /fonts

[font]
id=font_regular
style=default,LiberationSans-Regular.ttf,72,1
style=ja,unifont-10.0.06.ttf,72,1
style=ko,unifont-10.0.06.ttf,72,1
style=zh,unifont-10.0.06.ttf,72,1
style=zh_TW,unifont-10.0.06.ttf,72,1

[font]
id=font_captions
style=default,LiberationSans-Regular.ttf,72,1
style=ja,unifont-10.0.06.ttf,72,1
style=ko,unifont-10.0.06.ttf,72,1
style=zh,unifont-10.0.06.ttf,72,1
style=zh_TW,unifont-10.0.06.ttf,72,1

[font]
id=font_subtitles
style=default,LiberationSans-Regular.ttf,72,1
style=ja,unifont-10.0.06.ttf,72,1
style=ko,unifont-10.0.06.ttf,72,1
style=zh,unifont-10.0.06.ttf,72,1
style=zh_TW,unifont-10.0.06.ttf,72,1"

**Reply by dorkster** [reply]
I forgot to mention that the menu frame size in `engine/resolutions.txt` needs to be changed:

```
menu_frame_width=2560
menu_frame_height=1920
```

I also forgot that 2 properties for the config menu that we just use the engine defaults for; frame_offset and tab_offset. This should work if you put it in your `menus/config.txt`:

```
scrollpane=-24,-24,2464,1376
scrollpane_padding=320,160
frame_offset=44,32
tab_offset=12,0
```

You'll also probably want to increase the width of the text area in the horizontal list widget. In `engine/widget_settings.txt`:

```
[horizontal_list]
text_width=600
```

**Reply by WithinAmnesia** [reply]
![image](https://user-images.githubusercontent.com/86871191/222931585-3fdae49d-088d-49c2-b4c4-1c344407c2b8.png)
That seemed to do the trick; thank you Justin. I hope to get a 1080p to 4K working version of my alpha play test out sometime this month. I hope to have non-windows support included with no capital letters in the files hopefully with V.0.004!
